### ISSUE TITLE
Herwig7 installation in CMSSW_8_1_X not working

### ISSUE THREAD

**Issue by mharrend** [cms-sw/cmsdist#2571]
Dear all, 

with regards to the following merged PR #2477 , there are still some issues which make Herwig7 unusuable as I could find out by a test of the CMSSW_8_1_0_pre12 release:

1) The Herwig7 path is not automatically set via the use of 
`scram setup herwigpp`
, so that the Herwig7 binary is not found. This is only a minor issue.

2) If one sources the Herwig7 path manually via 
`source /cvmfs/cms.cern.ch/slc6_amd64_gcc530/external/herwigpp/7.0.2/etc/profile.d/init.sh`
Herwig7 could be invoked via
`Herwig read`

However, it gives the following error message:
`
Error: Could not find repository '/build/cmsbuild/auto-builds/CMSSW_8_1_0_pre12-slc6_amd64_gcc530/build/CMSSW_8_1_0_pre12-build/tmp/BUILDROOT/ca28c368082c42e94dd56564fd3bd483/opt/cmssw/slc6_amd64_gcc530/external/herwigpp/7.0.2/share/Herwig/HerwigDefaults.rpo'.
`
It seems that the Herwig paths are all set to the folder in which Herwig7 was built instead of pointing to its final installation folder: 
`
/cvmfs/cms.cern.ch/slc6_amd64_gcc530/external/herwigpp/7.0.2
`

In the herwigpp.spec I was stating
`
./configure $PLATF_CONF_OPTS \
            --with-thepeg=$THEPEG_ROOT \
            --with-fastjet=$FASTJET_ROOT \
            --with-gsl=$GSL_ROOT \
            --with-boost=$BOOST_ROOT \
        --with-madgraph=$MADGRAPH5AMCATNLO_ROOT \
            --with-openloops=$OPENLOOPS_ROOT \
--prefix=%i 
`
I would expect that
`
--prefix=%i
`
points by the variable i to the final installation folder.

Can you confirm that the variable points to the final installation folder instead of the build folder?
Is it expected that the error message above occurs?

Since Herwig7 sets many absolute paths during the make install step it is important that the prefix is pointing to its proper final location. Meaning that Herwig7 is at the moment absolutely useless due to point to the wrong folders.

Any ideas how we can solve the path issue?

Furthermore, the Herwig7 authors would be willing to help us, if you could provide a workflow similar to our build system which is independent of an LXPLUS installation or so and could be reproduced by them.

Thanks in advance for helping to get Herwig7 working in CMSSW.

**Comment by smuzaffar** [comment]
@mharrend , I am look in to it.
For path, I will update the herwigpp toolfile to add the bin directory.

**Comment by smuzaffar** [comment]
I guess if you source the init.sh files to get the herwigpp env then you also want to set HERWIGPATH=$HERWIGPP_ROOT/share/Herwig .  HERWIGPATH is set by scram tool file, so I guess, once we have bin directory in path then Herwig7 should work in cmssw env.

**Comment by mharrend** [comment]
@smuzzaffar Thanks for looking into this. 
Yes, I would also like to set the HERWIGPATH as you mentioned.

**Comment by smuzaffar** [comment]
@mharrend , Looks like at build time the herwig make system writes the build path in its sources e.g.

```
MatrixElement/Matchbox/Scales/MatchboxScale.cc:  makeFile<< "HERWIG=/var/build/muz/herwig/w/tmp/BUILDROOT/a25b3f229bd3328acd771d06c1f99c8e/opt/cmssw/slc6_amd64_gcc530/external/herwigpp/7.0.2-cms\n" ;
```

I am not sure if this is causing Herwig executable to look in the wrong directory for to file the repo file. I noticed that there is --repo command line arg which you can use to pass the rpo file e.g.

```
Herwig --repo $HERWIGPATH/HerwigDefaults.rpo  read
```

works once I have the Herwig bin directory in PATH. Can you make use of it?

**Comment by mharrend** [comment]
@smuzaffar Yes, that is exactly the problem during the build time Herwig writes in my places the build path in its sources. So, we have to make sure that Herwig is writing the final installation dir /cvmfs/... into its sources. In principle, this should happen if we set the final installation dir via --prefix.

But it seems that as a prefix the build path is set. Is this correct?

Unfortunately, your proposed solution via overloading Herwig

```
Herwig --repo $HERWIGPATH/HerwigDefaults.rpo
```

does not work in practice since the build path is used also for finding linked libraries and baked into the Herwig executable.

Can it be that we have to make sure that we make a stagged installation,
http://www.gnu.org/software/automake/manual/html_node/Staged-Installs.html
 e.g. like

```
~/amhello-1.0 % ./configure --prefix= builddir
…
~/amhello-1.0 % make
…
~/amhello-1.0 % make DESTDIR=/cvmfs...
```

**Comment by smuzaffar** [comment]
problem is that we do not know the DESTDIR at the time of build. We can say that we can hard code the /cvmfs path but then any one else installing it in different location will have problem. IBs do now use the same DESTDIR as releases.

Correct solution would be if herwig uses some environment (HERWIG_DIR??) variable which we can set via cmsenv? if veriable is not set then fallback to normal way of searching ?

**Comment by mharrend** [comment]
Okay, I believe there is no such variable. But I will contact the Herwig7 authors to ask them if they could include such an variable.

**Comment by iahmad-khan** [comment]
@mharrend Any update from Herwig7 authors , should we close this issue??

**Comment by mharrend** [comment]
@iahmad-khan No, I was meeting one of the Herwig authors on Sunday and should get a working proposal this week.
So please let this issue open for now. 
Thanks

**Comment by mharrend** [comment]
@smuzaffar  @iahmad-khan 
Dear CMS software experts,
I have now a proposal from the Herwig7 authors how we could fix the issue with the staged out installation:

1. We would build and compile Herwig7 as we do at the moment.
2. After the finished Herwig7 package is moved and installed on the /cvmfs system, we have to update a configuration file called HerwigDefaults.rpo which then contains the new paths.
This can be done via

```
.../Herwig init \
-L${pkglibdir} \
-L$(path to ThePEG) \
-i${pkgdatadir} \
-i${pkgdatadir}/snippets \
${pkgdatadir}/defaults/HerwigDefaults.in \
--repo=.../CMS-HerwigDefaults.rpo
```
in which `pkglibdir` is `/cvmfs/.../[Herwig7Installdir]/lib/Herwig` and `pkgdatadir` is `/cvmfs/.../[Herwig7Installdir]/share/Herwig`.

How can I define this in the herwig7.spec file?

Furthermore, I would also like to define the following alias:
```
alias Herwig="Herwig --repo=/.../.../.../CMS-HerwigDefaults.rpo"
```
So that also the Herwig7 standalone version would be working in CMSSW.

I would appreciate your help to get this added to the spec file or any other proper place.

Thanks,

Marco

**Comment by smuzaffar** [comment]
@mharrend , looking in to  /cvmfs/cms-ib.cern.ch/2016-48/slc6_amd64_gcc530/external/herwigpp/7.0.2-ghmdbj4/share/Herwig/HerwigDefaults.rpo file I see that all path points to /cvmfs install path. So basically we are already relocating this rpo file. Only thing which is not properly relocated is the ThePEG Path in it 
```
/cvmfs/cms-ib.cern.ch/2016-48/tmp/BUILDROOT/a5051c6e9d63baed7b1b1c96ad5fdc14/opt/cmssw/slc6_amd64_gcc530/external/thepeg/2.0.2-oenich/share/ThePEG
```
which still points to build directory. This can be fixed. So I do not think we need to run ``Herwig init ....``` right?

About setting the alias , there is nothing much we can do as scram does not handle alias. Also alias are not passed to to the scripts. I would suggest that you use Herwig --repo $HERWIGPATH/HerwigDefaults.rpo (HERWIGPATH env variable is set by scram).

**Comment by mharrend** [comment]
@smuzaffar Okay, yes, then it should be enough to fix the ThePEG path in the HerwigDefaults.rpo.
Can you please take care of this?

So, there is no way to overload the Herwig command that it refers to Herwig --repo ...?
It is not very important, but I would like to allow the user to use Herwig7 also as a standalone executable and of course, it would be nice if he does not has to add the --repo argument.

**Comment by smuzaffar** [comment]
@mharrend , sure we will fix the ThePEG issue.
I noticed that we have bin/Herwig++ which is a shell script and calls bin/Herwig. So how about we do the following
1. rename bin/Herwig to bin/Herwig-cms
2. create bin/Herwig shell script which calls bin/Herwig-cms --repo $HERWIGPATH/HerwigDefaults.rpo "$@"  (with some checks as Herwig does not like if --repo is passed twice)
3. bin/Herwig++ as usual calls  bin/Herwig

what do you say?

**Comment by mharrend** [comment]
@smuzaffar Sounds very good to me! Thanks for proposing :+1:

**Comment by smuzaffar** [comment]
@mharrend , this is fixed now in CMSSW_9_0_X_2016-11-29-2300. Can you please confirm that it works for you?

**Comment by smuzaffar** [comment]
please reopen this if you see any issues with herwig in lastest 90X IBs.
Open a new issue if this needs to be back ported to 81X

**Comment by mharrend** [comment]
Dear smuzzafar,

thank you very much it is working nicely!

Am Donnerstag, den 01.12.2016, 00:48 -0800 schrieb Malik Shahzad Muzaffar:
> Closed #2571.
>
